tiledb/sm/array_schema/attribute.h

```diff
@@ -237,9 +241,15 @@ class Attribute {
   void serialize(Serializer& serializer, uint32_t version) const;
 
   /**
-   * Sets the attribute number of values per cell. Note that if the attribute
-   * datatype is `ANY` this function returns an error, since `ANY` datatype
-   * must always be variable-sized.
+   * Sets the attribute number of values per cell.
+   *
+   * @throws AttributeException if `cell_val_num` is invalid
+   *
+   * A valid `cell_val_num` depends on the Attribute datatype and ordering.
+   * For `Datatype::ANY`, the only valid value is `constants::var_num`.
+   * If the attribute is unordered, then all other datatypes support any value.
+   * If the attribute is ordered, then an Attribute of `Datatype::STRING_ASCII`
+   * must have `constants::var_num`, and all other datatypes must have 1.
```

Comment:
Invariants such as these are statements about relationship between multiple member variables. As such, they should generally be listed in the documentation for the class as a whole.

In particular, `set_*` member functions that are C.41-hostile are very much not the right place.